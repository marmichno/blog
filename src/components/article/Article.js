import ArticlesCSS from './article.module.css';
import { useHistory } from "react-router-dom";
import {useDispatch} from 'react-redux';
import { chooseArticle } from '../../actions';
import React from 'react';

export const Article = ({id, title, paragraph}) => {

    const history = useHistory();
    const dispatch = useDispatch();

    const navigateTofullArticle = (e) => {

        const id = e.target.dataset.articleid;

        dispatch(chooseArticle(id));
        localStorage.setItem('articleId', id)

        const location = {
            pathname: '/article',
            state: {
                articleId:id
            }
        }
        history.push(location);
    }

    const favorites = (e) => {
        let target = e.target;
        const articleId = target.dataset.articleid;
        let favorites = localStorage.getItem('favorites');
        favorites = favorites ? favorites.split(',') : [];

        if(!target.classList.contains(`${ArticlesCSS['articleContainer__favourite--active']}`)){
            e.target.classList.add(`${ArticlesCSS['articleContainer__favourite--active']}`);
            favorites.push(articleId);
        }else{
            e.target.classList.remove(`${ArticlesCSS['articleContainer__favourite--active']}`);
            favorites = favorites.filter(value => value === articleId ? false : true);
        }

        localStorage.setItem('favorites', favorites.toString());
    }

    const checkIfFavorite = (id) => {
        const favorites = localStorage.getItem('favorites');
        if(favorites){
            return favorites.split(',').filter(value => parseInt(value) === id ? true : false).length > 0 ? true : false;
        }else{
            return false;
        }
    }

    return(
        <div data-testid="articleContainer" className={ArticlesCSS.articleContainer}>
            <span className={ArticlesCSS.articleContainer__image}></span>
            <div className={ArticlesCSS.articleContainer__textContainer}>
                <h2 className={ArticlesCSS.articleContainer__textContainer__header}>{title}</h2>
                <p className={ArticlesCSS.articleContainer__textContainer__published}>07/12/2021 | John Doe</p>
                <p className={ArticlesCSS.articleContainer__textContainer__paragraph}>{paragraph}</p>
                <button className={ArticlesCSS.articleContainer__textContainer__button} data-articleid={id} onClick={navigateTofullArticle}>Read more â¤´</button>
            </div>
            {
                checkIfFavorite(id) ? 
                <div className=
                {`${ArticlesCSS.articleContainer__favourite} ${ArticlesCSS['articleContainer__favourite--active']}`}
                onClick={favorites}
                data-testid="favorite"
                data-articleid={id}>
                </div>
                :
                <div data-testid="addToFavorite" className={ArticlesCSS.articleContainer__favourite} data-articleid={id} onClick={favorites}></div>
            }
        </div>
    )
}