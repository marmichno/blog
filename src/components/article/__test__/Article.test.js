import { createStore, applyMiddleware} from 'redux';
import { Provider } from 'react-redux';
import {Article} from '../Article';
import {fireEvent, render, waitFor} from '@testing-library/react';
import React from 'react';
import "@testing-library/jest-dom/extend-expect";
import TestRenderer from "react-test-renderer";

const store = createStore(() => [], {}, applyMiddleware());

const fakeLocalStorage = (() => {
    let store = {};
  
    return {
      getItem: (key) => {
        return store[key] || null;
      },
      setItem: (key, value) => {
        store[key] = value.toString();
      },
      removeItem: (key) => {
        delete store[key];
      },
      clear: () => {
        store = {};
      }
    };
  })();

beforeAll(() => {
    Object.defineProperty(window, 'localStorage', {
        value: fakeLocalStorage,
    });
});

afterEach(() => {
    jest.clearAllMocks();
});

const id = 1;
const title = "sunt aut facere repellat provident occaecati excepturi optio reprehenderit";
const paragraph = "quia et suscipit suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto";

test("Article renders with valid props passed", async () => {

    const {getAllByText} = render(<Provider store={store}><Article id={id} title={title} paragraph={paragraph}/></Provider>);

    await waitFor(() => {
        expect(getAllByText('sunt aut facere repellat provident occaecati excepturi optio reprehenderit'));
        expect(getAllByText('quia et suscipit suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto'));
    })
});

test("Container renders without props", async () => {

    const {getByTestId} = render(<Provider store={store}><Article id={undefined} title={undefined} paragraph={undefined}/></Provider>);

    await waitFor(() => {
        expect(getByTestId("articleContainer")).toBeTruthy();
    })
});

test("Favorite button recives class after click and article is added to local storage", async () => {

    const {getByTestId} = render(<Provider store={store}><Article id={id} title={title} paragraph={paragraph}/></Provider>);

    await waitFor(() => {
        fireEvent.click(getByTestId("addToFavorite"));
    });

    await waitFor(() => {
        expect(getByTestId("addToFavorite")).toHaveClass(`articleContainer__favourite--active`);
        expect(window.localStorage.getItem('favorites')).toEqual('1');
    })
});

test("Articles is removed from localstorage", async () => {

    const {getByTestId} = render(<Provider store={store}><Article id={id} title={title} paragraph={paragraph}/></Provider>);

    await waitFor(() => {
        fireEvent.click(getByTestId("favorite"));
    });

    await waitFor(() => {
        expect(window.localStorage.getItem('favorites')).toEqual(null);
    })
});

test("matches snapshot", () => {
    const tree = TestRenderer.create(<Provider store={store}><Article id={id} title={title} paragraph={paragraph}/></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  });