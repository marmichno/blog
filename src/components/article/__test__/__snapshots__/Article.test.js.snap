// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="articleContainer"
  data-testid="articleContainer"
>
  <span
    className="articleContainer__image"
  />
  <div
    className="articleContainer__textContainer"
  >
    <h2
      className="articleContainer__textContainer__header"
    >
      sunt aut facere repellat provident occaecati excepturi optio reprehenderit
    </h2>
    <p
      className="articleContainer__textContainer__published"
    >
      07/12/2021 | John Doe
    </p>
    <p
      className="articleContainer__textContainer__paragraph"
    >
      quia et suscipit suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto
    </p>
    <button
      className="articleContainer__textContainer__button"
      data-articleid={1}
      onClick={[Function]}
    >
      Read more â¤´
    </button>
  </div>
  <div
    className="articleContainer__favourite"
    data-articleid={1}
    data-testid="addToFavorite"
    onClick={[Function]}
  />
</div>
`;
